explorer/interpreter/resolve_unformed.cpp

```diff
@@ -0,0 +1,234 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "explorer/interpreter/resolve_unformed.h"
+
+#include <unordered_map>
+
+#include "common/check.h"
+#include "explorer/ast/ast.h"
+#include "explorer/ast/expression.h"
+#include "explorer/ast/pattern.h"
+#include "explorer/ast/static_scope.h"
+#include "explorer/common/error_builders.h"
+#include "explorer/common/nonnull.h"
+
+using llvm::cast;
+
+namespace Carbon {
+
+// Aggregate information about a ValueNodeView being analyzed.
+struct FlowFact {
+  bool is_formed;
+};
+
+// Traverses the sub-AST rooted at the given node, resolving the formed/unformed
+// states of local variables within it and updating the flow facts.
+static auto ResolveUnformed(
+    Nonnull<Expression*> expression,
+    std::unordered_map<ValueNodeView, FlowFact>& flow_facts,
+    bool set_formed = false) -> ErrorOr<Success>;
```

Comment:
Most of the calls to `ResolveUnformed` is about checking the flow facts instead of changing it so the default represents the majority of using cases of this function.

Reply:
When you say many, how many? Due to overloads, I can't count precisely -- if it's less than 10 using the default, I'd say to remove the default. From a simple typing perspective, you'd have to have at least 3 uses to break even, but I think there's also a benefit to understandability to explicitly pass parameters.

Reply:
Changed to explicitly passing parameters.